### ISSUE TITLE
Latency can potentially be reduced by 15ms.

### ISSUE THREAD

**Issue by trudnorx** [mumble-voip/mumble#3502]
Opus codec has 3 modes which are described roughly like this by the Opus documentation: optimize for "low bitrate speech" (SILK - high delay), "high bitrate speech / music" (CELT - medium delay), "low delay" (CELT - low delay).

From a quick review of the code it seems that Mumble always uses the "high delay" mode, except when the user manually changes "codec/opus/encoder/music" (bUseOpusMusicEncoding) which cannot be accessed from the UI. This seems pointless because from a few quick tests a) I did not notice any difference between high and medium delay at extremely low bitrates, b) any diffference that exists at all, even unnoticeable one, would probably disappear with >32kbps bitrate. I found some docs online ( https://wiki.hydrogenaud.io/index.php?title=Opus ) which recommend always using CELT with >32kb/s.

Therefore if other users can confirm a) I would recommend enabling bUseOpusMusicEncoding by default in all cases, otherwise it should be enabled by default when bitrate is higher than 32kb/s.

Regarding the low delay mode, this would save 7.5ms of latency, if it is used. It is simple to enable it as far as I know: Mumble simply needs to use OPUS_APPLICATION_RESTRICTED_LOWDELAY when it initializes the encoder.

Another 7.5ms could be saved if the user is allowed to set frame sizes as low as 2.5ms (which is only possible with certain of the modes described above). So a total of 15ms savings.

While the "low delay" and "2.5ms framesize" modes did seem to require higher bitrates to preserve quality, I think changes to allow these modes to be used should definitely be implemented, and indeed made default for certain bitrates, because in some quick tests, I was able to achieve good quality with 64kbit/s bitrate, low delay, 2.5ms framesize. If higher bitrate was needed I propose that Mumble allows selecting up to 510kbit/s when you enable "Advanced" checkbox in the options (the user clearly wanted to select Advanced mode, so must give him full freedom, IMO.)

**Comment by trudnorx** [comment]
Your post is a little confusing. You're saying there's a minor difference between "voip" and "audio" at 24k, and "no perceptible difference at all" between "audio" and "lowdelay"? In this case at which point is there no difference at all between "voip" and "audio"?

**Comment by Ytmndissue** [comment]
I'll be a bit more explicit for you:

**Use 56k.**

The difference is minor at 24k between the first two modes, VOIP and AUDIO as passed to FFMPEG.

The difference is gone at 56k.

The third command (I'm assuming) is supposed to represent how the 2.5ms frames sound? It is identical to the second command on all bit rates I hear, so 56k is again the bitrate to shoot for.

**Comment by Ytmndissue** [comment]
To be honest, I wouldn't even automatically consider the bitrate, the quality differences really are minor to where I don't even think switching is worth it. Just initialize it with OPUS_APPLICATION_RESTRICTED_LOWDELAY and call it a day.

**Comment by trudnorx** [comment]
I'm now working on adding options for low frame sizes.

So, actually, there's no way to configure the frame size at all currently. The "audio per packet" option in the Mumble settings does not actually configure the frame size itself at all (but then why does it only allow 10ms, 20ms, 40ms, 60ms, i.e. the acceptable Opus frame sizes below 10ms?). The frame size is always set to 10ms, no matter what.

How should this be resolved? Do we need two sliders or should the existing slider be changed to configure both things?

**Comment by Ytmndissue** [comment]
Weird, I'd say two sliders as that is less ambiguous.

**Comment by trudnorx** [comment]
Jitter buffer code needs to be changed and maybe other stuff too.

**Comment by trudnorx** [comment]
Good news, low delay Opus mode is working with aforementioned code but with buggy audio (you can make out words and stuff though), which proves the concept works, just need to fix it now
Low framesize works too, just also bugged audio

**Comment by Ytmndissue** [comment]
Any luck messing around with this?

**Comment by trudnorx** [comment]
No, I decided to take an indefinite break from this work when it was getting too complex. I was hoping that someone with more experience with the Mumble codebase could help. The problem is, while it's successfully making use of the low delay / low frame size modes (apart from one bug where you can't select certain combinations), the rest of the code isn't prepared to handle it, so it results in buggy audio. I think it has something to do with buffering, because it gets progressively worse the longer you've been running the program. Also, comments in the code indicate that the jitter buffer isn't prepared to handle such settings.

**Comment by trudnorx** [comment]
This issue is 2/3 implemented now.
The only remaining issue is allowing <10ms frame size down to 2.5ms

**Comment by trudnorx** [comment]
@davidebeatrici @Kissaki @mkrautz do you know anything about how Mumble audio code handles frame size? Would it readily support <10ms frame sizes?

**Comment by Kissaki** [comment]
I am not familiar with that

**Comment by Krzmbrzl** [comment]
Nah sorry. I don't know much about the audio code and given the nightmarish state of that part of the code I don't really want to start digging into it :see_no_evil: 

I think @davidebeatrici would probably be your best chance at this...

**Comment by trudnorx** [comment]
Ways to save latency

1. Allow choosing Opus frame size <10ms, for example 2.5ms. Currently, there's no option at all for this. This will decrease up to 15ms from roundtrip.
2. Allow using jitter buffer <10ms.
3. Allow using audio per packet <10ms.
4. #3503

**Comment by trudnorx** [comment]
@Krzmbrzl I think there's no hope, I've been trying to ask every dev on the project for over a year and they're either inactive, or don't know about the sound code... I want to focus on getting the lower Opus frame size to work it would result in some pretty considerable latency improvements... 

So far how I understand it is, Mumble initially creates an Opus encoder with opus_encoder_create(), one of the parameters is the frame size, and then, every time it feeds the encoder a bunch of data with opus_encode(), it specifies the frame size there too... but changing it in the parameter of the function wouldn't be good enough because the rest of the code, like the arrays that contain a certain amount of sound data, etc. need to be compatible with that frame size.

I tried to change it myself, but had problems like buggy audio, likely because I didn't change the rest of the code right, it seemed complicated, tangled... or maybe I tested it wrong. I think the comments said something about the way the jitter buffer is coded, it's not able to handle stuff like a lower frame size. So maybe that would have to be changed.

**Comment by felix91gr** [comment]
> I've been trying to ask every dev on the project for over a year and they're either inactive, or don't know about the sound code... I want to focus on getting the lower Opus frame size to work it would result in some pretty considerable latency improvements...

This is so sad :cry:
Mumble's latency is pretty great. Having it be lower still would probably open it to things not seen before, like remote music/chorus group practice.

**Comment by streaps** [comment]
The Mumble desktop client would need some other changes to be suitable for music/chorus group practice. There is also already some other software for low latency audio streaming, e.g.

https://github.com/elieserdejesus/JamTaba/
https://github.com/corrados/jamulus

**Comment by iainhallam** [comment]
> The Mumble desktop client would need some other changes to be suitable for music/chorus group practice. There is some hard coded audio processing that you don't want for music. There is also already some other software for low latency audio streaming...

Neither of those has an Android or iPhone app; in fact there's an interesting discussion on Sourceforge about using Mumble to connect to Jamulus sessions. Is there any way to turn off the processing relatively easily to reduce latency further?

**Comment by davidebeatrici** [comment]
#3323

**Comment by Chris2000SP** [comment]
If 2.5ms would be possible, then calculating 2.5ms audio processing  + 2x 2ms fiber latency to and from server + 2.5 for audio processing sums to 9ms. That could be a killer feature for Musicians. I would love that.

**Comment by davidebeatrici** [comment]
Keep in mind that there's also the audio backend's latency.

**Comment by Chris2000SP** [comment]
If you host the mumble server at home, you could get it lower if the people joining the server all has FttH.

**Comment by peterlvilim** [comment]
Thanks for work on this. Me and some friends used a local mumble server at a lan party to route our local chatter in the room through our headsets after finding this thread. Want latency as low as possible for that.

**Comment by Chris2000SP** [comment]
@peterlvilim Sorry, but today it is minimum 10ms on Output if you set the setting and 10ms on Input if you set this also. On top you get network latency.

Example:
Mumble has 20ms + Operating-systems Sound-Server Latency (Windows ASIO with 3 to 5ms OR Linux with Jack Soundserver or pipewire-jack also 3 to 5ms) + Network Latency you could get with FiberToTheHome and peer to peer round about 1ms on the same Provider and Region. This Equals to 24ms min and 26ms max now.

Example with a Patched Mumble:
2.5ms+2.5ms on Mumble + OS 3-5ms + Network 1ms FTTH PtP = 9ms.

9 milliseconds would be great for Playing Music.

**Comment by peterlvilim** [comment]
Cool! Do patches up thread still work? I might try building from source for
me and my friends

**Comment by Chris2000SP** [comment]
@peterlvilim  If you can write a Patch that would be great.

**Comment by MaxNarr** [comment]
This looks really interesting. Any news on this ?

**Comment by Krzmbrzl** [comment]
Nope. But things like this should definitely be considered for libmumble which is intended to eventually provide the core functionality for the desktop client.

**Comment by Chris2000SP** [comment]
Ahm, i found that claude has some good coding skills and i wanted to ask if i could give it a try with my little coding skills with that Feature here? I mean claude seemed to have a solution to it though. I make a gist file of the output but i have a geraman answering to that:
https://gist.github.com/Chris2000SP/bcdc99e85f21dbe4df2d9e3b502d403a

Sorry i did that in German.

Could that work?

**Comment by Krzmbrzl** [comment]
It might. The main problem here is that Mumble's audio code is really ancient and has a lot of implicit assumptions about sample rates, frame sizes etc. baked into it. This makes it really difficult to figure out what may break when changing some of these assumptions.

Given how central audio is for Mumble, I would probably not feel to good about making any changes as fundamental as this before having rigorous (unit) test cases in place that will ensure things remain working. The problem here is that we currently don't have such any such tests.

For this reason, I believe that it will probably be easiest to postpone features such as this one to the reimplementation in libmumble. That is intended to encapsulate the core Mumble protocol (including audio en-/decoding) and be used in Mumble in the long term. That being said, the reimplementation has kinda stalled due to lack of time.

However, if someone wants to give this a shot in the current code base, I think it would have to be in these steps (ideally as separate PRs so they can be reviewed, tested and merged individually though 1 and 2 could go togerher)
1. Refactor audio en-/decoding into standalone classes (decoupled from audio input and output)
2. Write unit tests ensuring things work as expected (including simulated packet loss)
3. Refactor audio input and output to get rid of implicit assumptions about sample rate etc. This will means that all relevant parameters will have to be stored in variables explicitly and then everything as to use those variables.
4. Once the above has been verified to work as intended, changes to supported audio quality and frame sizes can be made (along with corresponding unit tests).
